### PR TITLE
First go at AltiVec RAID-Z

### PR DESCRIPTION
**Pull request by rdolbeau**
Implements the RAID-Z function using AltiVec SIMD.
This is basically the NEON code translated to AltiVec.

Note that the 'fletcher' algorithm requires 64-bits
operations, and the initial implementations of AltiVec
(PPC74xx a.k.a. G4, PPC970 a.k.a. G5) only has up to
32-bits operations, so no 'fletcher'.

Signed-off-by: Romain Dolbeau <romain.dolbeau@european-processor-initiative.eu>

### Motivation and Context

Performance only, on a limited amount of hardware...

### Description

This add AltiVec (PowerPC SIMD) support for RAID-Z SIMD.

However :

1) This has only received limited testing on my G4, G5 & QEMU
2) This has ~~only~~ been tested on a big-endian 64 bits PPC (ppc64) & BE 32 bits PPC (ppc)
3) ~~This should be tested on little-endian 64 bits PPC (ppc64el)~~ I've also tested in on Debian ppc64el in QEMU, it passes ztest & raidz_test
4) The testing code for userland is the old 'try altivec and catch sigill' method, I don't know of anything better
5) ~~The testing code for kernel is disabled, has it seems to be GPL-only on my kernel and I don't know how to properly detect AltiVec in-kernel (catching SIGILL is probably not an option there...)~~ This checks the MSR 'Vec' bit is set when AltiVec is enabled
6) ~~Seems that adding -maltivec is crashing every non-ppc arch, I'm not sure how to properly add the option to the file that needs it (the compiler doesn't seem to want to deal with AltiVec asm without -maltivec)~~ ~~Fixed, but Makefile implementation might not be very clean.~~ Makefile was updated to use the symbols introduced with #9848 
7) no 32 bits ztest, userland for BE is 32 bits and ztest crashes at start-up:

```
dolbeau@powermacg5:~/zfs$ ./cmd/ztest/ztest loading concrete vdev 0, metaslab 14 of 15 ...
error: Pool 'ztest' has encountered an uncorrectable I/O failure and the failure mode property for this pool is set to panic.
/home/dolbeau/zfs/cmd/ztest/.libs/ztest(+0x9058)[0x789058]
linux-vdso32.so.1(__kernel_sigtramp32+0x0)[0x100424]
/lib/powerpc-linux-gnu/libc.so.6(gsignal+0xdc)[0xf7a6e4bc]
/lib/powerpc-linux-gnu/libc.so.6(abort+0x144)[0xf7a55978]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(+0x40314)[0x370314]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(vcmn_err+0x0)[0x3703b0]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(+0x17a42c)[0x4aa42c]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(+0x17c07c)[0x4ac07c]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(zio_execute+0xd8)[0x4a7f78]
/home/dolbeau/zfs/lib/libzpool/.libs/libzpool.so.2(+0x41758)[0x371758]
/lib/powerpc-linux-gnu/libpthread.so.0(+0x891c)[0x1d891c]
/lib/powerpc-linux-gnu/libc.so.6(clone+0x60)[0xf7b3d438]
child died with signal 6
```
However, ztest seems OK in QEMU on ppc64el with 64 bits userland.

8) ~~disable_kernel_altivec() was apparently introduced around kernel 4.5 so fails to compile on e.g. 3.16~~ disable_kernel_altivec() is only used for kernel >= 4.5

Performance (on G5):
```
(dolbeau)powermacg5:~/zfs> cat /proc/spl/kstat/zfs/vdev_raidz_bench 
18 0 0x01 -1 0 1022640300308 1320961408741                                                                                                                                                                                                                                     
implementation   gen_p           gen_pq          gen_pqr         rec_p           rec_q           rec_r           rec_pq          rec_pr          rec_qr          rec_pqr                                                                                                       
original         160563916       43900725        27402800        482443964       67385690        11490536        22298016        3420657         3158063         2105862                                                                                                       
scalar           694230532       199962624       86366216        607205930       174317751       116252651       62091132        53510959        42755587        30793098                                                                                                      
powerpc_altivec  1027742844      488162276       222745665       890691014       493951582       329358008       213486266       185015426       156249270       107822494                                                                                                     
fastest          powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec powerpc_altivec       
```

### How Has This Been Tested?

Only tested on a BE PPC64 970MP "G5" running 4.19, both raidz_test & trying a pool.
Also on a BE PPC 7455 "G4" running 5.3 (and to a limited extent 3.16).
Also on a LE POWER9 in QEMU, running 4.19, ztest & raiz_test.

ztest was _not_ run on BE systems, see above.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ x] I have updated the documentation accordingly.
- [ x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ x] All new and existing tests passed.
- [ x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
@rdolbeau I should have access to a couple ppc64el systems I can give this a spin on.  Though I might not be able to get to it for a little bit.

**Comment by rdolbeau** [comment]
@behlendorf Thanks, and there's no hurry. Even it if all architectures works, there's still the issue of detecting AltiVec in-kernel, I'm still not sure how to do that. And there is even some current PPC64 w/o AltiVec - the e5500 core in the new AmigaOne X5000 is one of them - so enabling AltiVec all the time isn't a good option.

Though the real fun will start with the variable-length SIMD ISA like Arm's SVE and RISC-V's V :-)

Question - do the ppc & ppc64 buildbot do any test that could validate the code, or do they really just build?

**Comment by behlendorf** [comment]
> do the ppc & ppc64 buildbot do any test that could validate the code

They really only verify the compilation and don't perform any testing.

**Comment by Ornias1993** [comment]
I'm going to be a bit harsh here, but is there any use for this?
Spending time getting it to work, without a test suite... On a architecture that is by all means depricated and still requires significant work to get working in any decent fasion?

**Comment by rdolbeau** [comment]
@Ornias1993 It's a legitimate question to ask :-) I'll try to answer...

a) Big-endian PPC (32, 64) is obsolete for desktop/laptop/servers, but there's still some Linux-supported hardware out there, and it's still in use in the embedded market (and don't tell the Amiga crowd their current processor of choice is obsolete ;-) ). Also, the older hardware needs the performance boost the most;

b) Little-endian PPC (64) is very much alive at IBM, with POWER8 and POWER9 out there and POWER10 announced - and as far as I understand, standard AltiVec code should work on those VSX-enabled systems (they might use a bit more parallelism that what is in this patch, if someone wants to donate me a Blackbird mainboard with CPU and cooler, I'll make sure to check & tune for POWER9 ;-) ) .

Also, this doesn't require "significant work to get working", as ZFS is already working on those architectures, and the SIMD infrastructure is common to all of them. It's only a bit of assembly, and detecting the availability of the ISA.

And also, why the port to this hardware? "Because it's there" :-)

Cordially,

**Comment by Ornias1993** [comment]
a) its obsolete. period. sue me.
b) Okey, thats a good argument you got there, missed the POWER10 announcement, thanks! :)
c) (-ish) I was mistaken in my understanding it was to get ZFS working, thanks for clearing that up :)

**Comment by rdolbeau** [comment]
@behlendorf This should work on ppc64el now (QEMU only for me), and I've added some in-kernel detection (and updated my description accordingly).

**Comment by rdolbeau** [comment]
rebased & rechecked.

File: lib/libzpool/Makefile.am [hunk 1]
```diff
@@ -202,3 +203,6 @@ libzpool_la_LIBADD += $(ZLIB) -ldl
 libzpool_la_LDFLAGS = -pthread -version-info 2:0:0
 
 EXTRA_DIST = $(USER_C)
+
+vdev_raidz_math_powerpc_altivec.$(OBJEXT): CFLAGS += $(if $(findstring powerpc,$(target_cpu)),-maltivec,)
+vdev_raidz_math_powerpc_altivec.l$(OBJEXT): CFLAGS += $(if $(findstring powerpc,$(target_cpu)),-maltivec,)
```
Comment (behlendorf) [inline lib/libzpool/Makefile.am]: Actually, one issue.  This code isn't portable and generates the following warning when included in the `Makefile.am`.

```
lib/libzpool/Makefile.am:207: warning: if $(findstring powerpc,$(target_cpu: non-POSIX variable name
lib/libzpool/Makefile.am:207: (probably a GNU make extension)
lib/libzpool/Makefile.am:208: warning: if $(findstring powerpc,$(target_cpu: non-POSIX variable name
lib/libzpool/Makefile.am:208: (probably a GNU make extension)
```

Comment (rdolbeau) [reply lib/libzpool/Makefile.am]: @behlendorf Thanks for the testing & the numbers:

As for the Makefile.am, yes, it's far from ideal (point 6 in my original comments). My Makefile-fu is failing me here, I don't know how to do that 'cleanly' ('that' being: add the -maltivec option when compiling the AltiVec file on powerpc, but not on other platforms). Any suggestion on how to improve that welcome!

Comment (behlendorf) [reply lib/libzpool/Makefile.am]: I've opened PR #9848 with some proposed cleanup to facilitate the needed Makefile additions.  With those changes applied we can make the following Makefile changes.  Assuming there aren't any major downsides to adding `-maltivec` to more then just the new `vdev_raidz_math_powerpc_altivec.c` file.  If so, we can restrict it.

lib/libzpool/Makefile.am
```
# Required by vdev_raidz_math_powerpc_altivec.o
if TARGET_CPU_POWERPC
AM_CFLAGS += -maltivec
endif
```

For the kernel side using `CONFIG_ALTIVEC` to enable this feels more correct.

module/zfs/Makefile.in
```
# Required by vdev_raidz_math_powerpc_altivec.o
cflags-$(CONFIG_ALTIVEC) += -maltivec

...

$(MODULE)-$(CONFIG_ALTIVEC) += vdev_raidz_math_powerpc_altivec.o
```

**Comment by rdolbeau** [comment]
@behlendorf Now that #9848 is merged, I'll try to rebase/upgrade and use the new CPU handling.

**Comment by rdolbeau** [comment]
Rebased, updated to use #9848, rechecked with raidz_test on 32BE/64BE/64LE.

File: module/zfs/vdev_raidz_math_powerpc_altivec.c [hunk 2]
```diff
@@ -0,0 +1,4337 @@
...
+	return (kfpu_allowed()) && zfs_altivec_available();
+}
+
+const raidz_impl_ops_t vdev_raidz_powerpc_altivec_impl = {
+	.init = NULL,
+	.fini = NULL,
+	.gen = RAIDZ_GEN_METHODS(powerpc_altivec),
+	.rec = RAIDZ_REC_METHODS(powerpc_altivec),
+	.is_supported = &raidz_will_powerpc_altivec_work,
+	.name = "powerpc_altivec"
```
Comment (behlendorf) [inline module/zfs/vdev_raidz_math_powerpc_altivec.c]: "powerpc_altivec" is right at the 16 character limit which causes `EINVAL` to be returned when trying to set it with `echo powerpc_altivec >/sys/module/zfs/parameters/zfs_vdev_raidz_impl".  Increasing `RAIDZ_IMPL_NAME_MAX` from 16 to 20 resolves the issue.  Alternately we could shorten the name to "altivec".

Comment (rdolbeau) [reply module/zfs/vdev_raidz_math_powerpc_altivec.c]: a) Will fix preemption ASAP

b) For BE, I did test with raidz_test, loading modules and some zpool/zfs operations, but ztest itself always crashes on my (32 bits userland) BE systems

c) I count only 15 for powerpc_altivec... I'll push the limit to 20, as I think we should keep the $arch_$simd nomenclature for clarity.

**Comment by rdolbeau** [comment]
@behlendorf I still need to merge the commits, I'll do that ASAP

**Comment by behlendorf** [comment]
@rdolbeau sounds good.  Alternately, I can squash them when merging if you prefer.

**Comment by rdolbeau** [comment]
@behlendorf if you can squash while merging it's OK for me.

BTW - I couldn't get the pre-emption issue message in syslog before the patch, weirdly. I might not have run operations long-running enough for the issue to show up though :-(

**Pull request merged**